tests/Filter/StringFilterTest.php

```diff
@@ -29,110 +29,182 @@ public function testEmpty(): void
         $filter->filter($builder, 'alias', 'field', '');
+        $filter->filter($builder, 'alias', 'field', []);
 
         $this->assertSame([], $builder->query);
         $this->assertFalse($filter->isActive());
     }
 
-    public function testNullValue(): void
+    public function getValues(): iterable
+    {
+        return [
+            'filter by normal value' => ['asd', false],
+            'not filter by empty string' => ['', false],
+            'filter by empty string' => ['', true],
+            'not filter by null' => [null, false],
+            'filter by null' => [null, true],
+        ];
+    }
+
+    /**
+     * @dataProvider getValues
+     */
+    public function testDefaultType(?string $value, bool $allowEmpty): void
     {
         $filter = new StringFilter();
-        $filter->initialize('field_name');
+        $filter->initialize('field_name', ['allow_empty' => $allowEmpty]);
 
         $builder = new ProxyQuery(new QueryBuilder());
         $this->assertSame([], $builder->query);
 
-        $filter->filter($builder, 'alias', 'field', ['value' => null, 'type' => StringOperatorType::TYPE_EQUAL]);
-        $this->assertFalse($filter->isActive());
+        $filter->filter($builder, 'alias', 'field', ['value' => $value, 'type' => null]);
+
+        if ($value) {
+            $this->assertSame(['alias.field LIKE :field_name_0'], $builder->query);
+            $this->assertSame(['field_name_0' => sprintf('%%%s%%', $value)], $builder->parameters);
+            $this->assertTrue($filter->isActive());
+        } else {
+            $this->assertSame([], $builder->query);
+            $this->assertFalse($filter->isActive());
+        }
     }
 
-    public function testContains(): void
+    /**
+     * @dataProvider getValues
+     */
+    public function testContains(?string $value, bool $allowEmpty): void
     {
         $filter = new StringFilter();
-        $filter->initialize('field_name');
+        $filter->initialize('field_name', ['allow_empty' => $allowEmpty]);
 
         $builder = new ProxyQuery(new QueryBuilder());
         $this->assertSame([], $builder->query);
 
-        $filter->filter($builder, 'alias', 'field', ['value' => 'asd', 'type' => StringOperatorType::TYPE_CONTAINS]);
-        $this->assertSame(['alias.field LIKE :field_name_0'], $builder->query);
-        $this->assertSame(['field_name_0' => '%asd%'], $builder->parameters);
-
-        $builder = new ProxyQuery(new QueryBuilder());
-        $this->assertSame([], $builder->query);
+        $filter->filter($builder, 'alias', 'field', ['value' => $value, 'type' => StringOperatorType::TYPE_CONTAINS]);
 
-        $filter->filter($builder, 'alias', 'field', ['value' => 'asd', 'type' => null]);
-        $this->assertSame(['alias.field LIKE :field_name_0'], $builder->query);
-        $this->assertSame(['field_name_0' => '%asd%'], $builder->parameters);
-        $this->assertTrue($filter->isActive());
+        if ($value) {
```

Comment:
IMO, we should make this condition explicit, otherwise values like `0` or `'0'` will be interpreted as false. Indeed, these values should be added to the list of the provided values.

Reply:
If we try to use `0`, we get error:

> Argument 1 passed to Sonata\DoctrineORMAdminBundle\Tests\Filter\StringFilterTest::testNotEquals() must be of the type string or null, int given

Reply:
I missed that you were declaring `?string` for the value in the test method.
I think this is ok only if the filter really can't handle other types (like the integer in my comment).

Reply:
I would say that the string filter is made to deal with string only.

Reply:
>I would say that the string filter is made to deal with string only.

I agree, that's the intention. But since the method is public, there is no check preventing other types and the provided value is being casted as string, I think a test covering these cases could help to avoid unintended behaviors.
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/80b37e6336b3fb9293da35fde06a4b1eef8e1835/src/Filter/StringFilter.php#L55

Reply:
I add data type casting

Reply:
Calls to `ltrim()` and `rtrim()` were already performing the casting.

Reply:
No, because a string from the `StringFilter::trim()` method comes to them. The typecast had to be done before calling `StringFilter::trim()`.